### PR TITLE
Feature: change retreive_token_metadata to fallback to OpenSea

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: api/notd/manager.py [hunk 1]
```diff
@@ -181,6 +182,7 @@ async def retrieve_ui_data(self, startDate: datetime.datetime, endDate: datetime
             limit=1
         )
         transactionCount = await self.retriever.get_transaction_count(startDate=startDate,endDate=endDate)
+        logging.info(highestPricedTokenTransfers[0])
```
Comment (krishan711) [inline api/notd/manager.py:185]: remove logs

File: api/notd/manager.py [hunk 2]
```diff
@@ -281,7 +283,17 @@ async def retreive_token_metadata(self, registryAddress: str, tokenId: str) -> R
...
         if len(tokenMetadatas) == 0:
-            raise NotFoundException()
+            tokenMetadatas = []
+            cacheKey = f'{registryAddress}:{tokenId}'
+            if cacheKey in self._tokenCache:
+                return self._tokenCache[cacheKey]
+            try:
+                registryToken = await self.openseaClient.retreive_registry_token(registryAddress=registryAddress, tokenId=tokenId)
+            except NotFoundException:
+                registryToken = await self.tokenClient.retreive_registry_token(registryAddress=registryAddress, tokenId=tokenId)
```
Comment (krishan711) [inline api/notd/manager.py:293]: no need for this - its the same thing our database has

Comment (Femi-Ogunkola) [reply api/notd/manager.py:293]: So for this should I remove the try block entirely?

Comment (krishan711) [reply api/notd/manager.py:293]: yep

File: api/notd/manager.py [hunk 3]
```diff
@@ -281,7 +282,13 @@ async def retreive_token_metadata(self, registryAddress: str, tokenId: str) -> R
             ], limit=1,
         )
         if len(tokenMetadatas) == 0:
-            raise NotFoundException()
+            tokenMetadatas = []
```
Comment (krishan711) [inline api/notd/manager.py:285]: i dont think you need this

File: api/notd/manager.py [hunk 4]
```diff
@@ -281,7 +282,13 @@ async def retreive_token_metadata(self, registryAddress: str, tokenId: str) -> R
...
+            cacheKey = f'{registryAddress}:{tokenId}'
+            if cacheKey in self._tokenCache:
+                return self._tokenCache[cacheKey]
```
Comment (krishan711) [inline api/notd/manager.py:288]: dont return here, set it to registryToken and skip the two lines below

Comment (krishan711) [reply api/notd/manager.py:288]: in fact this is really bad - its returning something of the wrong type!

File: api/notd/manager.py (hunk 3 above)
Comment (Femi-Ogunkola) [reply api/notd/manager.py:285]: But how do we keep  the return statement the same  the tokenMetadatas[0]

Comment (krishan711) [reply api/notd/manager.py:285]: isnt `tokenMetadatas` already an empty list at this point?

Comment (Femi-Ogunkola) [reply api/notd/manager.py:285]: Ohh

**Review by krishan711 — APPROVED** [review]

**Review by krishan711 — APPROVED** [review]

**Pull request merged**